keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
+```
+
+2. RFC7838 HTTP Alternative Services
+
+This specification defines a new concept in HTTP, "Alternative Services", that allows an origin server to nominate additional means of interacting with it on the network.
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, this requires that the users of this feature belong to the cluster network. TODO define a way of filtering the headers in the answer, per example, by source IP, certificate name, ... 
+
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
+```
+
+The Alternative Services RFC defines options to [implement caching](https://datatracker.ietf.org/doc/html/rfc7838#section-3.1),
+however, it doesn't seem necessary because:
+
+- kubernetes control-plane is not a high dynamic environment.
+- apiserver already has another mechanisms to warn clients using http codes.
+- if the alt-service has gone you will have a connection error, the client will expire that entry from the cache and will try another alt-svc.
+
+### Client side HA
+
+Client-go creates a layer on top of the golang net/http library abstracting the communications against the apiservers using [Requests](https://github.com/kubernetes/client-go/blob/master/rest/request.go).
+
+The golang net/http library allows to modify a request in different points:
+- RoundTripper/Transport
+- Dialer
+- DNS dialer
```

Comment:
Yeah, just a note that this point can be used to identify e.g. multiple A records.

Reply:
the problem is to generate the records without adding dependencies, if we remove the LB to add an external DNS server the win is 0, actually ,you can do it today, using an external DNS server I mean, there is no need to modify anything

Reply:
I disagree that there is no need to modify anything, the client currently doesn't do anything useful if there happen to be multiple A records.

Whether an LB or DNS system is easier to run, I am not sure.

Reply:
> I disagree that there is no need to modify anything, the client currently doesn't do anything useful if there happen to be multiple A records.

It does, however, the golang implementation of the http client doesn't refresh the DNS, but I know how to fix that if you want

> Whether an LB or DNS system is easier to run, I am not sure.

The point is to remove external dependencies, an LB or a DNS system are point of failures managed by different teams or hided by external services.

Reply:
> It does, however, the golang implementation of the http client doesn't refresh the DNS, but I know how to fix that if you want

Honestly it seems like a reasonable addition to the default go client, at least as options (periodic re-resolve DNS; load balance requests over known IPs). I'm not sure if they'd accept such a change though. I have done something similar in the past and I agree it's fairly annoying.

> The point is to remove external dependencies, an LB or a DNS system are point of failures managed by different teams or hided by external services.

One still needs an LB for the first call under the design proposed here. I don't think this solves the problem.

Reply:
> Honestly it seems like a reasonable addition to the default go client, at least as options (periodic re-resolve DNS; load balance requests over known IPs). I'm not sure if they'd accept such a change though. I have done something similar in the past and I agree it's fairly annoying.

the trick is to call CloseIdleConnections(), if there is no connection in flight next call will resolve, some person from Zalando that hangs out in sig-network have this working ... but indeed is a hack.

> One still needs an LB for the first call under the design proposed here. I don't think this solves the problem.

one needs an apiserver for the first call, the LB just bounces the request, I think we are running through this in several comments and I can't see where is our disconnection, you can also use a DNS record for bootsrapping, the first resolution always work

or  poor man solution :)

> It is though? The way I understood it was that bootstrap can be done using AlternativeHosts

Reply:
> the trick is to call CloseIdleConnections(), if there is no connection in flight next call will resolve, some person from Zalando that hangs out in sig-network have this working ... but indeed is a hack.

This won't work for controllers which have many requests in flight basically all the time.

> I think we are running through this in several comments and I can't see where is our disconnection

We have a complexity budget (I know, I know, bear with me...); k8s devs, cluster users, cluster admins all pay with their souls for complexity. A strange DNS replacement (such as described in this KEP) is a LOT of complexity (mostly for k8s devs and cluster admins), so we need to get a lot of benefit, or remove a lot of complexity somewhere else. Unfortunately, in this design many/most cluster admins would still not be able to turn off their LBs. So, this is adding complexity for everyone (just having the option is complexity, even if you leave it off), while benefiting a minority (say, < 50%). That is not a good tradeoff for a new feature.